### PR TITLE
perf: optimize construction & search for HNSW

### PR DESCRIPTION
**Pull request by BubbleCal**
on 8.5M rows dataset

before this:
ef_c=150, build=2473.328s, recall=0.94

after this:
ef_c=150, build=1947.985s, recall=1
ef_c=50, build=1235.757s, recall=0.98

changes:
- pruning if the number of neighbors exceeds `m_max`
- skip pruning if no enough candidates
- select only 1 entry point for the next level while inserting
- avoid re-calculating the distances between `query` and `candidates`

### CODE REVIEW

File: rust/lance-index/src/vector/graph.rs [hunk 1]
```diff
@@ -171,20 +171,16 @@ pub trait Graph {
...
+    let mut visited: HashSet<_> = start.iter().map(|node| node.id).collect();
     let dist_calc = dist_calc.unwrap_or_else(|| graph.storage().dist_calculator(query).into());
-    let mut candidates: BinaryHeap<Reverse<OrderedNode>> = dist_calc
-        .distance(start)
-        .iter()
-        .zip(start)
-        .map(|(&dist, id)| Reverse((dist.into(), *id).into()))
-        .collect();
+    let mut candidates: BinaryHeap<Reverse<OrderedNode>> =
+        start.iter().cloned().map(Reverse).collect();
```
Comment (chebbyChefNEQ) [inline rust/lance-index/src/vector/graph.rs:183]: nit: you could do `collection::<BinaryHeap<_>>`

File: rust/lance-index/src/vector/graph/builder.rs [hunk 2]
```diff
@@ -1,167 +1,161 @@
...
-            .into_iter()
-            .map(|n| n.id)
-            .collect::<Vec<_>>();
-        Self {
-            id: node.id,
-            neighbors,
-        }
-    }
-}
+// impl From<&GraphBuilderNode> for GraphNode<u32> {
```
Comment (chebbyChefNEQ) [inline rust/lance-index/src/vector/graph/builder.rs:32]: should we just delete this?

File: rust/lance-index/src/vector/graph.rs [hunk 3]
```diff
@@ -234,7 +230,8 @@ pub(super) fn beam_search(
             }
         }
     }
-    Ok(results)
+
+    Ok(results.into_sorted_vec())
```
Comment (eddyxu) [inline rust/lance-index/src/vector/graph.rs:234]: TIL

File: rust/lance-index/src/vector/graph/builder.rs (hunk 2 above)
Comment (eddyxu) [reply rust/lance-index/src/vector/graph/builder.rs:32]: So here you changed `Levels<Node>` to `Node<Levels>`?
Curious what is the perf benefit of that.

But lets delete the commented out code nonetheless.

File: rust/lance-index/src/vector/hnsw.rs [hunk 4]
```diff
@@ -127,29 +125,29 @@ impl HnswLevel {
...
-            schema.into(),
-            vec![
-                Arc::new(vector_id_builder.finish()),
-                Arc::new(neighbours_builder.finish()),
-            ],
-        )?;
-
-        Ok(Self::new(batch, vectors))
-    }
+    // fn from_builder(builder: &BuilderLevelView, vectors: Arc<dyn VectorStorage>) -> Result<Self> {
```
Comment (eddyxu) [inline rust/lance-index/src/vector/hnsw.rs:128]: Delete these code?

File: rust/lance-index/src/vector/hnsw/builder.rs [hunk 5]
```diff
@@ -183,92 +197,155 @@ impl HNSWBuilder {
...
-            graphs,
+            self,
             self.entry_point,
             self.vectors.metric_type(),
             self.params.use_select_heuristic,
         ))
     }
 }
+
+pub struct BuilderLevelView<'a> {
```
Comment (eddyxu) [inline rust/lance-index/src/vector/hnsw/builder.rs:309]: Could you add a comment about what is this. what is the logical relationship between this and HnsfBuilder.

Also should this be `private` or `pub(crate)`?
Can you rename it to `HnswBuilderLevelView` or `HnswLevelView`?

**Review by eddyxu — APPROVED** [review]
Some nits.

**Pull request merged**